### PR TITLE
Variational consistent postprocessing

### PR DESCRIPTION
**Pull request by kmokstad**
This contains some framework for doing so-called variational consistent postprocessing, where an explicit extraction function (dual solution field) is used to calculate sectional forces.

### CODE REVIEW

**Comment by kmokstad** [comment]
jenkins build this with downstreams ifem-elasticity=85 please

**Comment by kmokstad** [comment]
jenkins build this with downstreams ifem-elasticity=85 please

**Comment by kmokstad** [comment]
jenkins build this with downstreams ifem-elasticity=85 please

**Comment by akva2** [comment]
You need to add the new header to the list of installation.

**Comment by kmokstad** [comment]
jenkins build this with downstreams ifem-elasticity=85 please

**Comment by kmokstad** [comment]
jenkins build this with downstreams ifem-elasticity=85 please

**Comment by kmokstad** [comment]
jenkins build this with downstreams ifem-elasticity=85 please

**Comment by kmokstad** [comment]
The last commit fixes #347

**Comment by kmokstad** [comment]
jenkins build this with downstreams ifem-elasticity=85 please

**Comment by kmokstad** [comment]
jenkins build this with downstreams ifem-elasticity=85 please

**Comment by kmokstad** [comment]
jenkins build this with downstreams ifem-elasticity=85 please

**Comment by kmokstad** [comment]
jenkins build this with downstreams ifem-elasticity=85 please

**Comment by kmokstad** [comment]
jenkins build this with downstreams ifem-elasticity=85 please

**Comment by kmokstad** [comment]
jenkins build this with downstreams ifem-elasticity=85 please

File: src/SIM/FunctionSum.C [hunk 1]
```diff
@@ -0,0 +1,56 @@
+// $Id$
+//==============================================================================
+//!
+//! \file FunctionSum.C
+//!
+//! \date Apr 16 2019
+//!
+//! \author Knut Morten Okstad / SINTEF
+//!
+//! \brief Unary spatial function as a sun of other spatial functions.
```
Comment (akva2) [inline src/SIM/FunctionSum.C]: suM

Comment (kmokstad) [reply src/SIM/FunctionSum.C]: Fixed

**Comment by kmokstad** [comment]
jenkins build this with downstreams ifem-elasticity=85 please

File: src/ASM/ASMbase.h [hunk 2]
```diff
@@ -613,6 +613,14 @@ class ASMbase
   //!
   //! \note The implementation of this method is placed in GlbL2projector.C
   bool L2projection(Matrix& fVals, FunctionBase* function, double t = 0.0);
+  //! \brief Projects explicit functions using a continuous global L2-fit.
+  //! \param[out] fVals Control point values of the functions
+  //! \param[in] functions The functions to project
+  //! \param[in] t Current time
+  //!
+  //! \note The implementation of this method is placed in GlbL2projector.C
+  bool L2projection(const std::vector<Matrix*>& fVals,
```
Comment (akva2) [inline src/ASM/ASMbase.h]: doxy mismatch ((function vs functions).

File: src/ASM/GlbL2projector.h [hunk 3]
```diff
@@ -107,8 +112,11 @@ class GlbL2 : public Integrand
   //! \brief Solves the projection equation system and evaluates nodal values.
   //! \param[out] sField Nodal/control-point values of the projected results.
   bool solve(Matrix& sField);
+  //! \brief Solves the projection equation system and evaluates nodal values.
+  //! \param[out] sField Nodal/control-point values of the projected results.
+  bool solve(const std::vector<Matrix*>& sFields);
```
Comment (akva2) [inline src/ASM/GlbL2projector.h]: doxy mismatch (sFields vs sField).

File: src/SIM/FunctionSum.C [hunk 4]
```diff
@@ -59,8 +59,17 @@ bool FunctionSum::initPatch (size_t idx)
...
+  for (size_t i = 0; i < comps.size(); i++)
+    if (comps[i].second > 0.0)
+      sum.add(comps[i].first->getValue(X),comps[i].second);
+    else if (i == 0)
+      sum = comps[i].first->getValue(X);
+    else // find the max value
+    {
+      std::vector<double> val = comps[i].first->getValue(X);
+      for (size_t i = 0; i < val.size(); i++)
+        if (val[i] > sum[i]) sum[i] = val[i];
```
Comment (akva2) [inline src/SIM/FunctionSum.C]: no fabs intended?

Comment (kmokstad) [reply src/SIM/FunctionSum.C]: no

**Comment by kmokstad** [comment]
jenkins build this with downstreams ifem-elasticity=85 please

**Comment by kmokstad** [comment]
jenkins build this with downstreams ifem-elasticity=85 ifem-poisson=35 ifem-advectiondiffusion=30 ifem-darcy=20 please

**Comment by kmokstad** [comment]
jenkins build this with downstreams ifem-elasticity=85 please

**Pull request merged**